### DISCUSSION TITLE
Select All rows from all pages with controlled pagination

### DISCUSSION THREAD

**Discussion by ioanarucareanu** [TanStack/table#3513] [Q&A]
Hi! 
I have been using react-table and implemented back-end pagination, select functionality and it all works great. However, I have been trying to wrap my head around implementing also the functionality of selecting all the rows in all the pages (for e.g like in Gmail, having a link/button allowing you to select/unselect the total number of items) and all my efforts were in vain until now. 
Any lead on this? How could I accomplish a Select All from all pages functionality, since my data in not fully loaded at the point of selection? Any help/ suggestion is welcome :) Thank you in advance!

**Comment by tevinch** [comment]
You can keep the selection intent outside the table's loaded-row map, using two modes:

```js
// Individually selected IDs, kept across pages:
{ scope: 'catalog-v1:Guides', mode: 'include', ids: ['item-002'] }

// Every eligible match of that query, except these IDs:
{ scope: 'catalog-v1:Guides', mode: 'exclude', ids: ['item-004'] }
```

Selecting all switches to `exclude` with an empty array; unchecking a row adds its ID to the exclusions. For each page, project that descriptor onto only the page's eligible IDs. Keep the page checkbox separate from the “all matching results” control, so selecting another page does not confuse its mixed state.

The backend must understand the descriptor too: resolve the scope to its own matching eligible dataset, then apply the inclusion or exclusion IDs. A frontend flag alone cannot make an endpoint process rows it has not been told how to select. Reset when filters or the dataset/eligibility revision change; paging and sorting can preserve the same scope.

This follows the [maintainer's guidance to own the all-selection logic](https://github.com/TanStack/table/issues/4781#issuecomment-5192037966). I put an original [MIT module and runnable read-only example](https://github.com/tevinch/data-shape-kit/tree/main/examples/cross-page-selection) together for this pattern. The module has no runtime dependencies; the full example targets TanStack Table 9.2.4 and includes a small local server, page controls, exclusions and result preview. It is an application pattern, so older react-table versions need their own table wiring.
